homeassistant/components/motionblinds_ble/cover.py

```diff
@@ -0,0 +1,717 @@
+        device_class=CoverDeviceClass.BLIND
+    ),
+    MotionBlindType.DOUBLE_ROLLER.value: MotionCoverEntityDescription(),
+    MotionBlindType.CURTAIN.value: MotionCoverEntityDescription(
+        device_class=CoverDeviceClass.CURTAIN
+    ),
+    MotionBlindType.VERTICAL.value: MotionCoverEntityDescription(
+        device_class=CoverDeviceClass.CURTAIN, icon=ICON_VERTICAL_BLIND
+    ),
+}
+
+
+async def async_setup_entry(
+    hass: HomeAssistant, entry: ConfigEntry, async_add_entities: AddEntitiesCallback
+) -> None:
+    """Set up blind based on a config entry."""
+
+    blind_type = BLIND_TO_ENTITY_TYPE[entry.data[CONF_BLIND_TYPE]]
+    blind = blind_type(entry)
+
+    hass.data[DOMAIN][entry.entry_id] = blind
+
+    async_add_entities([blind])
+
+
+class GenericBlind(CoverEntity):
+    """Representation of a blind."""
+
+    device_address: str
+    device_rssi: int | None = None
+    _device: MotionDevice
+    _attr_connection_type: MotionConnectionType = MotionConnectionType.DISCONNECTED
+    _running_type: MotionRunningType | None = None
+
+    _use_status_position_update_ui: bool = False
+
+    _battery_callback: Callable[[int | None], None] | None = None
+    _speed_callback: Callable[[MotionSpeedLevel | None], None] | None = None
+    _connection_callback: Callable[[MotionConnectionType | None], None] | None = None
+    _signal_strength_callback: Callable[[int | None], None] | None = None
+
+    def __init__(self, entry: ConfigEntry) -> None:
+        """Initialize the blind."""
+        _LOGGER.info(
+            "(%s) Setting up %s cover entity (%s)",
+            entry.data[CONF_MAC_CODE],
+            entry.data[CONF_BLIND_TYPE],
+            BLIND_TO_ENTITY_TYPE[entry.data[CONF_BLIND_TYPE]].__name__,
+        )
+        super().__init__()
+        self.entity_description = COVER_TYPES[entry.data[CONF_BLIND_TYPE]]
+        self.config_entry: ConfigEntry = entry
+        self.device_address: str = entry.data[CONF_ADDRESS]
+        self._attr_name: str = ENTITY_NAME.format(mac_code=entry.data[CONF_MAC_CODE])
+        self._attr_unique_id: str = entry.data[CONF_ADDRESS]
+        self._attr_device_info: DeviceInfo = DeviceInfo(
+            identifiers={(DOMAIN, entry.data[CONF_MAC_CODE])},
+            manufacturer=MANUFACTURER,
+            name=self._attr_name,
+        )
+        self._attr_is_closed: bool | None = None
+        self._attr_is_opening: bool | None = None
+        self._attr_is_closing: bool | None = None
+        self._attr_should_poll: bool = False
+
+    async def async_added_to_hass(self) -> None:
+        """Run when entity about to be added."""
+        ble_device = (
+            async_ble_device_from_address(self.hass, self.device_address)
+            if self.device_address
+            else None
+        )
+        self._device = MotionDevice(
+            self.device_address, ble_device, device_name=self._attr_name
+        )
```

Comment:
This probably does not matter, but this could also be done when the config entry is setup and the bluetooth device is first discovered. That way you wouldn't have to search the device again using `async_ble_device_from_address`. The `MotionDevice` instance could then be stored in `hass.data`.

Reply:
Initially I didn't know how to pass data from config flow to an entity, but I could have indeed used this. However, right now I use `hass.data` to save the cover entity, which is then used by the other platforms to register callbacks. `hass.data[DOMAIN][entry.entry_id] = blind`

Reply:
Sure, I think it's fine either way. If you should change your mind about using `hass.data` you could of course always assign a dictionary to store multiple different objects.